src/lib/optimizer/strategy/join_to_predicate_rewrite_rule.cpp

```diff
@@ -0,0 +1,200 @@
+  }
+}
+
+bool JoinToPredicateRewriteRule::_check_rewrite_validity(const std::shared_ptr<JoinNode>& join_node,
+                                                         std::optional<LQPInputSide>& removable_side,
+                                                         std::shared_ptr<PredicateNode>& valid_predicate) const {
+  std::shared_ptr<AbstractLQPNode> removable_subtree = nullptr;
+
+  if (removable_side) {
+    removable_subtree = join_node->input(*removable_side);
+  } else {
+    // we know the join can only be a semi join in this case, where the right input is the one removed
+    auto r = LQPInputSide::Right;
+    removable_side = r;
+    removable_subtree = join_node->right_input();
+  }
+
+  std::shared_ptr<BinaryPredicateExpression> join_predicate = nullptr;
+  const auto& join_predicates = join_node->join_predicates();
+  for (auto& predicate : join_predicates) {
+    join_predicate = std::dynamic_pointer_cast<BinaryPredicateExpression>(predicate);
+    if (join_predicate) {
+      break;
+    }
+  }
+
+  DebugAssert(join_predicate, "A Join must have at least one BinaryPredicateExpression.");
+
+  std::shared_ptr<AbstractExpression> exchangable_column_expr = nullptr;
+
+  if (expression_evaluable_on_lqp(join_predicate->left_operand(), *removable_subtree)) {
+    exchangable_column_expr = join_predicate->left_operand();
+  } else if (expression_evaluable_on_lqp(join_predicate->right_operand(), *removable_subtree)) {
+    exchangable_column_expr = join_predicate->right_operand();
+  }
+
+  DebugAssert(exchangable_column_expr,
+              "Neither column of the join predicate could be evaluated on the removable input.");
+
+  // Check for uniqueness
+  auto testable_expressions = ExpressionUnorderedSet{};
+  testable_expressions.insert(exchangable_column_expr);
+
+  if (!removable_subtree->has_matching_unique_constraint(testable_expressions)) {
+    return false;
+  }
+
+  // Now, we look for a predicate that can be used inside the substituting table scan node.
+  visit_lqp(removable_subtree, [&removable_subtree, &valid_predicate](auto& current_node) {
+    if (current_node->type == LQPNodeType::Union) return LQPVisitation::DoNotVisitInputs;
+    else if (current_node->type != LQPNodeType::Predicate) return LQPVisitation::VisitInputs;
+
+    const auto candidate = std::static_pointer_cast<PredicateNode>(current_node);
+    const auto candidate_exp = std::dynamic_pointer_cast<BinaryPredicateExpression>(candidate->predicate());
+
+    if (!candidate_exp) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    DebugAssert(candidate_exp, "Need to get a predicate with a BinaryPredicateExpression.");
+
+    // Only predicates that filter using equals cond. on a column with a constant value are of use to our optimization
+    // these conditions have the potential (given filtered column is a UCC) to single out a maximum of one result tuple
+    if (candidate_exp->predicate_condition != PredicateCondition::Equals)
```

Comment:
I just thought about this when reviewing the discovery PR. It schould not be a problem to rewrite joins with any condition. A join `A.a < B.b` can safely be rewritten to a scan `A.a < <subquery_result>`. We inly have to encure that the predicate condition might have to be flipped when we swap the sides of the predicate (i.e., `A.a` was on the right side of the predicate before and your rewrite places it on the left side).

Reply:
Are you sure about this? I mean, in a table scan, I will never have any tuples of the input relation duplicated, the output cardinality is not greater than the input cardinality. With a join on a less than predicate, I can have duplicate rows of one input relation, even if we filter on a UCC. Wouldn't the two result relations differ in that case?

Reply:
Yes, you're right. My fault.